File: 0133-clone-graph/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) return node;

        // value = index and start with 1
        Queue<Node> q = new LinkedList<>();
        q.add(node);
        Map<Node, Node> visited = new HashMap<>();
        visited.put(node, new Node(node.val, new ArrayList()));

        while (!q.isEmpty()) {
            Node cur = q.remove();
            for (Node nb : cur.neighbors) {
                if (!visited.containsKey(nb)) {
                    visited.put(nb, new Node(nb.val, new ArrayList()));
                    q.add(nb);
                }
                visited.get(cur).neighbors.add(visited.get(nb));
            }
        }
        return visited.get(node);
    }
}

File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
from collections import deque

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if node is None:
            return node

        q = deque([node])
        # key: original node | val: clone node
        has_cloned = {node: Node(node.val)}

        while q:
            cur = q.popleft()
            for nb in cur.neighbors:
                if nb not in has_cloned:
                    has_cloned[nb] = Node(nb.val)
                    q.append(nb)
                has_cloned[cur].neighbors.append(has_cloned[nb])

        return has_cloned[node]